### PR TITLE
Higher-level multiboot interface

### PR DESCRIPTION
**Pull request by hugelgupf**
- users shouldn't have to figure out what trampoline to use -- it's an implementation detail (make the multiboot package figure it out on its own)
 - use boot.MultibootImage instead of multiboot.Multiboot -- the multiboot package should be internal.

### CODE REVIEW

File: cmds/kexec/kexec_linux.go [hunk 1]
```diff
@@ -138,7 +86,7 @@ func main() {
 	var kernelpath string
```
Comment (n-canter) [inline cmds/kexec/kexec_linux.go]: Wouldn't it be clearer if you write smth like:

```
if opts.load {
    kernelpath := flag.Arg(0)
    var image boot.OSImage
    if err := multiboot.Probe(kernelpath); err == nil {
       image = &boot.MultibootImage{...}
    } else {
       image = &boot.LinuxImage{...}
   }
   image.Load()
}
```

File: pkg/multiboot/multiboot.go [hunk 2]
```diff
@@ -89,27 +90,60 @@ func Probe(file string) error {
...
+// load loads and parses multiboot information from m.file.
+func (m *multiboot) load(debug bool) error {
 	log.Printf("Parsing file %v", m.file)
 	b, err := readFile(m.file)
 	if err != nil {
 		return err
 	}
 	kernel := kernelReader{buf: b}
-	log.Println("Parsing Multiboot Header")
+	log.Println("Parsing multiboot Header")
```
Comment (n-canter) [inline pkg/multiboot/multiboot.go]: s/Header/header

File: pkg/multiboot/multiboot.go [hunk 3]
```diff
@@ -129,9 +163,9 @@ func (m *Multiboot) Load(debug bool) error {
 		return fmt.Errorf("Error parsing memory map: %v", err)
 	}
 
-	log.Printf("Preparing Multiboot Info")
+	log.Printf("Preparing multiboot Info")
```
Comment (n-canter) [inline pkg/multiboot/multiboot.go]: s/Info/info

File: pkg/multiboot/multiboot.go [hunk 4]
```diff
@@ -247,7 +281,7 @@ func min(a, b uint32) uint32 {
 	return b
 }
 
-func (m *Multiboot) newMultibootInfo() (*infoWrapper, error) {
+func (m *multiboot) newmultibootInfo() (*infoWrapper, error) {
```
Comment (n-canter) [inline pkg/multiboot/multiboot.go]: s/newmultibootInfo/newMultibootInfo

sed based refactoring? :-)

File: pkg/multiboot/multiboot.go [hunk 5]
```diff
@@ -286,20 +320,20 @@ func (m *Multiboot) newMultibootInfo() (*infoWrapper, error) {
 
 // Segments returns kexec.Segments, where all the multiboot related
 // information is stored.
-func (m Multiboot) Segments() []kexec.Segment {
+func (m multiboot) Segments() []kexec.Segment {
```
Comment (n-canter) [inline pkg/multiboot/multiboot.go]: Do we still need this method exproted?

File: pkg/multiboot/multiboot.go [hunk 6]
```diff
@@ -76,7 +77,7 @@ type MemoryMap struct {
 	Type uint32
 }
```
Comment (n-canter) [inline pkg/multiboot/multiboot.go]: Do we need both MemoryMap and MemoryMaps exported?

File: pkg/multiboot/description.go [hunk 7]
```diff
@@ -36,7 +36,7 @@ type Description struct {
 
 // Description returns string representation of
 // multiboot information.
-func (m Multiboot) Description() (string, error) {
+func (m multiboot) Description() (string, error) {
```
Comment (n-canter) [inline pkg/multiboot/description.go]: I think we can unexport this method, it is called only if the kernel loaded in debug/verbose mode.

File: pkg/multiboot/multiboot.go (hunk 6 above)
Comment (hugelgupf) [reply pkg/multiboot/multiboot.go]: not MemoryMaps, but MemoryMap, because multiboot.Description uses MemoryMap

File: cmds/kexec/kexec_linux.go [hunk 8]
```diff
@@ -123,39 +71,33 @@ func main() {
...
-			log.Printf("%s is a multiboot v1 kernel.", kernelpath)
-			l = mboot{
-				debug:   opts.debug,
-				modules: opts.modules,
+			image = &boot.MultibootImage{
+				Modules: opts.modules,
+				Path:    kernelpath,
+				Cmdline: newCmdline,
+			}
+		} else {
```
Comment (n-canter) [inline cmds/kexec/kexec_linux.go]: one more condition
else if err == multiboot.ErrFlagsNotSupported {
   log.Fatal(err)	
}

**Pull request merged**